File: app_tarefas_compartilhadas/lib/models/historico_status.dart
import 'tarefa.dart';
import 'usuario.dart';

class HistoricoStatus {
  final int? id;
  final Tarefa? tarefa;
  final Usuario? usuario;
  final String? statusAnterior;
  final String? statusNovo;
  final String? dataAlteracao;

  HistoricoStatus({
    this.id,
    this.tarefa,
    this.usuario,
    this.statusAnterior,
    this.statusNovo,
    this.dataAlteracao,
  });

  factory HistoricoStatus.fromJson(Map<String, dynamic> json) {
    return HistoricoStatus(
      id: json['id'],
      tarefa: json['tarefa'] != null
          ? Tarefa.fromJson(json['tarefa'])
          : null,
      usuario: json['usuario'] != null
          ? Usuario.fromJson(json['usuario'])
          : null,
      statusAnterior: json['statusAnterior'],
      statusNovo: json['statusNovo'],
      dataAlteracao: json['dataAlteracao'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'tarefa': tarefa?.toJson(),
      'usuario': usuario?.toJson(),
      'statusAnterior': statusAnterior,
      'statusNovo': statusNovo,
      'dataAlteracao': dataAlteracao,
    };
  }
}

File: tarefas-compartilhadas/src/main/java/br/edu/tarefascompartilhadas/model/HistoricoStatus.java
package br.edu.tarefascompartilhadas.model;

import br.edu.tarefascompartilhadas.enums.StatusTarefa;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Entity
public class HistoricoStatus {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "tarefa_id", nullable = false)
    private Tarefa tarefa;

    @ManyToOne
    @JoinColumn(name = "usuario_id", nullable = false)
    private Usuario usuario;

    @Enumerated(EnumType.STRING)
    private StatusTarefa statusAnterior;

    @Enumerated(EnumType.STRING)
    private StatusTarefa statusNovo;

    private LocalDateTime dataAlteracao;

    @PrePersist
    public void prePersist() {
        if (dataAlteracao == null) {
            dataAlteracao = LocalDateTime.now();
        }
    }
}
